### PR TITLE
feat(webserver): Reload repository cache when scheduler job finishes

### PR DESCRIPTION
**Pull request by boxbeam**
TAB-520

### CODE REVIEW

File: ee/tabby-webserver/src/cron/mod.rs [hunk 1]
```diff
@@ -21,9 +28,11 @@ pub async fn run_cron(
     auth: Arc<dyn AuthenticationService>,
     job: Arc<dyn JobService>,
     worker: Arc<dyn WorkerService>,
+    repositories: Arc<RepositoryCache>,
     local_port: u16,
 ) {
     let mut jobs = vec![];
+    let (send, mut receive) = broadcast::channel::<SchedulerJobCompleteEvent>(2);
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/cron/mod.rs:35]: seems this is not used anywhere for listening the event?

Comment (wsxiaoys) [inline ee/tabby-webserver/src/cron/mod.rs:35]: how about:

1. define a CronEvents struct
```
struct CronEvents {
  scheduler_job_complete: channel
}
```

and listen the event as part of RepositoryCache::new?

Comment (boxbeam) [reply ee/tabby-webserver/src/cron/mod.rs:35]: Yes, I could do that. I was thinking I'd probably make the event an enum if we need to add more types of them, though.

File: ee/tabby-webserver/src/cron/mod.rs [hunk 2]
```diff
@@ -3,11 +3,39 @@ mod scheduler;
...
 use tokio_cron_scheduler::{Job, JobScheduler};
 use tracing::error;
 
 use crate::schema::{auth::AuthenticationService, job::JobService, worker::WorkerService};
 
+#[derive(Clone)]
+pub(crate) struct SchedulerJobCompleteEvent;
+
+pub(crate) struct CronEvents {
+    pub scheduler_job_complete: Receiver<SchedulerJobCompleteEvent>,
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/cron/mod.rs:17]:
```suggestion
    pub scheduler_job_succeeded: Receiver<()>,
```

File: ee/tabby-webserver/src/cron/mod.rs [hunk 3]
```diff
@@ -3,11 +3,39 @@ mod scheduler;
...
+}
+
+pub(crate) fn start_listener<E, F, Fut>(recv: &Receiver<E>, f: F)
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/cron/mod.rs:20]: how about implement this as a trait to `Receiver`?

File: ee/tabby-webserver/src/cron/mod.rs [hunk 4]
```diff
@@ -22,34 +50,41 @@ pub async fn run_cron(
     job: Arc<dyn JobService>,
     worker: Arc<dyn WorkerService>,
     local_port: u16,
-) {
+) -> Option<CronEvents> {
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/cron/mod.rs:53]:
```suggestion
) -> CronEvents {
```

File: ee/tabby-webserver/src/handler.rs [hunk 5]
```diff
@@ -28,10 +29,11 @@ pub async fn attach_webserver(
...
 ) -> (Router, Router) {
     let ctx = create_service_locator(logger, code, is_chat_enabled).await;
-    cron::run_cron(ctx.auth(), ctx.job(), ctx.worker(), local_port).await;
-
     let repository_cache = Arc::new(RepositoryCache::new_initialized(ctx.repository()).await);
-    repository_cache.start_reload_job().await;
+    let events = cron::run_cron(ctx.auth(), ctx.job(), ctx.worker(), local_port).await;
+    if let Some(events) = events {
+        repository_cache.start_reload_listener(&events);
+    }
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/handler.rs:36]:
```suggestion
    let events = cron::run_cron(ctx.auth(), ctx.job(), ctx.worker(), local_port).await;
    let repository_cache = RepositoryCache::new_initialized(ctx.repository(), events. scheduler_job_succeeded).await
```

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

File: ee/tabby-webserver/src/cron/mod.rs (hunk 4 above)
Comment (boxbeam) [reply ee/tabby-webserver/src/cron/mod.rs:53]: This function can fail in various ways and return early

File: ee/tabby-webserver/src/handler.rs (hunk 5 above)
Comment (boxbeam) [reply ee/tabby-webserver/src/handler.rs:36]: Unless we want to be using `.unwrap()` on `run_cron`, we can't do this

File: ee/tabby-webserver/src/cron/mod.rs (hunk 4 above)
Comment (wsxiaoys) [reply ee/tabby-webserver/src/cron/mod.rs:53]: Then let's use Result<CronEvents>

File: ee/tabby-webserver/src/handler.rs (hunk 5 above)
Comment (wsxiaoys) [reply ee/tabby-webserver/src/handler.rs:36]: `.await?`

File: ee/tabby-webserver/src/repositories/resolve.rs [hunk 6]
```diff
@@ -66,7 +71,7 @@ impl RepositoryCache {
 
     pub fn start_reload_listener(self: &Arc<Self>, events: &CronEvents) {
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/repositories/resolve.rs:72]: private?

File: ee/tabby-webserver/src/handler.rs [hunk 7]
```diff
@@ -29,11 +29,9 @@ pub async fn attach_webserver(
     local_port: u16,
 ) -> (Router, Router) {
     let ctx = create_service_locator(logger, code, is_chat_enabled).await;
-    let repository_cache = Arc::new(RepositoryCache::new_initialized(ctx.repository()).await);
     let events = cron::run_cron(ctx.auth(), ctx.job(), ctx.worker(), local_port).await;
-    if let Some(events) = events {
-        repository_cache.start_reload_listener(&events);
-    }
+    let repository_cache = RepositoryCache::new_initialized(ctx.repository(), &events).await;
+    repository_cache.start_reload_listener(&events);
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/handler.rs:34]:
```suggestion
```

File: ee/tabby-webserver/src/repositories/resolve.rs [hunk 8]
```diff
@@ -37,18 +39,23 @@ impl std::fmt::Debug for RepositoryCache {
...
         if let Err(e) = cache.reload().await {
             error!("Failed to load repositories: {e}");
         };
+        let cache = Arc::new(cache);
+        cache.start_reload_listener(events);
         cache
     }
 
-    async fn reload(&self) -> Result<()> {
+    pub async fn reload(&self) -> Result<()> {
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/repositories/resolve.rs:58]:
```suggestion
   async fn reload(&self) -> Result<()> {
```

File: ee/tabby-webserver/src/cron/mod.rs [hunk 9]
```diff
@@ -3,11 +3,45 @@ mod scheduler;
...
+pub trait StartListener<E> {
+    fn start_listener<F, Fut>(&self, handler: F)
+    where
+        F: Fn(E) -> Fut + Send + 'static,
+        Fut: Future + Send,
+        E: Clone + Send + 'static;
+}
+
+impl<E> StartListener<E> for Receiver<E> {
+    fn start_listener<F, Fut>(&self, handler: F)
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/cron/mod.rs:25]: Please add unit test

**Review by wsxiaoys — COMMENTED** [review]
Otherwise LGTM

**Pull request merged**